### ISSUE TITLE
User Experience -- redesign/rethink the command line interface

### ISSUE THREAD

**Issue by kytrinyx** [exercism/discussions#35]
I'd like to think about this from scratch, aiming for a complete overhaul (CLI v3.x).

I think that for each command we should think through:
- what are the basic use cases?
- what are reasonable defaults?
- what is the person going to want to do right after calling it?
- what information does the person need after calling it?

Here are some thoughts:

**Top level commands**
- `upgrade` - The current implementation is good, and user experience is probably fine.
- `debug` - I think we have the information we need here.
- `configure` - Mostly I think configuration is OK, but there's one big change that I want to make to the configuration, and that is that the CLI should only talk to one API, not two. If we have several services, then let one proxy to the other. We should include versioning in the base URL, so that we can experiment with alternate APIs if we like. E.g.

```
exercism configure --api=http://exercism.io/api/v3/
```

Or, maybe we experiment with rewriting the API in clojure:

```
exercism configure --api=http://api.exercism.io/api/v1/
```

I think we should continue to use the configured exercism directory plus the location of the files to determine which track and problem we're submitting. E.g. if your exercism directory is `/home/alice/exercism` and you are submitting `/home/alice/exercism/java/hello-world/src/main/whatever/super/nested/thingy.java` then it will use chop off the exercism dir, and then use the next two segments to figure out the track and problem: `java` and `hello-world`.

**Basic workflow**

A user will generally be working on a specific track (or some specific tracks).

I think that the `skip` command is fine. It would only work if you specify a full TRACK_ID and EXERCISE_ID.

I think that it would be very, very useful if we had some commands that helped get things started:
- `exercism setup TRACK_ID` - this would walk you through downloading any of the necessary dependencies that you need in order to work on stuff in that language. There might also be track-level files that we could deliver (e.g. gradle stuff, maybe, for java)
- `exercism tutorial TRACK_ID` - this would be an interactive tutorial that walks you through solving a problem in exercism (probably hello-world). You'd only submit hello world if you're working through the tutorial.

Then the usual usage would be to fetch and submit.

`fetch` - I think this is working well right now.
- `exercism fetch` - fetch the next exercise in all the tracks that the user has available in the current directory (that way if someone doesn't want to work on a track they can move that directory or rename it).
- `exercism fetch TRACK_ID` - fetch the next exercise (based on what you've submitted before)
- `exercism fetch TRACK_ID EXERCISE_ID` - fetch the exercise that is specified

`submit` - I like the multi-file submit, but I think that we can add a better default when you don't pass any arguments after the track name.
- `exercism submit` - detect which track it is, and submit all the files in the current directory plus all the subdirectories, including the test file... the site should be able to identify what are tests, and put those last (or even hide them by default, with the option of showing them).
- `exercism submit TRACK_ID` - same as above, complain if TRACK_ID doesn't match the track of the directory you are in.
- `exercism submit TRACK_ID file1 [file2, ...] - submit specific files, assuming that they're in the current directory.

It kind of feels like we should be able to always detect the track, but I hate the inconsistency of allowing `exercism submit file1 [file2, ...]`.

We also have a `restore` command which downloads all the solutions you've submitted. This would work much better if we have people submit the entire current directory.

**Context**

When people are between exercises, I think there are a lot of things you might wonder about.
- What are the available tracks?
- What tracks have I done work on?
- What exercises are available in a given track?
- What is the URL for my TRACK_ID EXERCISE_ID exercise? (we currently have `exercism open`)
- How many exercises do I have left in a given track?

`tracks` and `list` can be combined:
- `exercism tracks` - list tracks, maybe along with a link to the track page on exercism.
- `exercism tracks TRACK_ID` - list available problems in a track.

`status` - is hopelessly broken.

@Tonkpils observed in https://github.com/exercism/discussions/issues/13: _We should think about what status means in each context (global, track, problem). In each of those contexts we could scope down what the command displays in a much better way. That is unless we want to just keep it a one specific context?_

In other words,  we need to think through what it means to call:
- `exercism status`
- `exercism status TRACK_ID`
- `exercism status TRACK_ID EXERCISE_ID`

**Reviewing**

`download` When reviewing other people's code, I think that the idea of being able to quickly download their exercise is a good one, but right now we kind of stick files in a super awkward place. Let's think about how to do this better.

**Comment by Tonkpils** [comment]
I like where this is going and I agree 💯 % with all your points. This will give us the opportunity to clean up the codebase and enhance the test suit for the CLI. 

## Submit

It may be a bit of a stretch but if we know the files that have to be submitted. There's usually not a need to submit files outside of the ones that the exercise is expecting. We could boil it down to 

```
exercism submit
exercism submit TRACK_ID EXERCISE_ID
```

`submit TRACK_ID` looks like the odd one out since we don't know which exercise they want to submit unless we make the assumption and have a way of knowing their latest unsubmitted exercise?

## Download

Is the purpose of downloading the file to review it works or actually review the source code? Is this a feature that a lot of people use? If the concern is to review other people's code, it would make sense to spend more time enhancing the interface for the code review on the website. If the idea is to do a side-by-side comparison of a submission with code you have locally maybe we can approach this by just adding a download button on the web? Most likely they are already on the website and have to copy and paste the submission link anyway.

It would be good to get some feedback on this feature and get some sense on the usage of it so we can make a better decision on whether to keep this or not. 

## Status

### TBD

**Comment by kytrinyx** [comment]
> submit TRACK_ID looks like the odd one out since we don't know which exercise they want to submit unless we make the assumption and have a way of knowing their latest unsubmitted exercise?

Yeah, it's the odd one out. I feel like we could just default to submitting the current dir, if it's a subdirectory of $TRACK_ID/, just for the heck of supporting it, but that the documentation would suggest using one of the other two.

> Is the purpose of downloading the file to review it works or actually review the source code?

I sometimes want to run the code, and other times I want to try out a change before I point something out or make a comment. The problem now is that we merge their solution with the current test suite which (a) might not be the version they solved against, and (b) has skipped/pending tests in a lot of tracks.

**Comment by kytrinyx** [comment]
> Is [download] a feature that a lot of people use?

That's a very good question. I have no idea. I probably use it about 2 or 3 percent of the time. Not a whole lot. I still think that we should have a way to make it easy-ish to do.

**Comment by kytrinyx** [comment]
> It would be good to get some feedback on this feature and get some sense on the usage of it so we can make a better decision on whether to keep this or not.

Yeah, agreed. I'll stick it in this weeks newsletter, and tweet about it. I think we need to fix the dashboard so I can start asking people questions about stuff like this :)

**Comment by kotp** [comment]
I use the download pretty often.  Before I realized it was there I kept overwriting my solution with theirs, to explore.  Having the download makes this process much easier.

**Comment by kytrinyx** [comment]
@exercism/cli-committers I realized that I forgot to tag the whole team on this issue.

**Comment by kytrinyx** [comment]
Ah, jeez. I forgot I ever opened this, and opened a new one in https://github.com/exercism/discussions/issues/129

The discussion there has all of these ideas, and more. I'm closing this one in favor of the new one.
